File: WheelsCodingChallenge/Controller/ListVC.m
//
//  ListVC.m
//  WheelsCodingChallenge
//
//  Created by Mike Saradeth on 4/4/19.
//  Copyright © 2019 Mike Saradeth. All rights reserved.
//

#import "ListVC.h"
#import "ListTableCell.h"
#import "AddUserVC.h"

@interface ListVC ()

@end

@implementation ListVC
@synthesize tableView;
@synthesize items;

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupVC];
    
    #pragma mark - Load data from server
    [StackApiService loadData:^(NSMutableArray *items) {
        self.items = items;
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.tableView reloadData];
        });
    }];
}

#pragma mark - Setup VC
- (void)setupVC {
    self.navigationController.navigationBar.prefersLargeTitles = true;
    items = [[NSMutableArray alloc] init];
    tableView.dataSource = self;
    [tableView registerNib:[UINib nibWithNibName:@"ListTableCell" bundle:nil] forCellReuseIdentifier:@"Cell"];
}


#pragma mark - Add new user button action
- (IBAction)addButtonAction:(id)sender {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"StackUserList" bundle:nil];
    AddUserVC *vc = [storyboard instantiateViewControllerWithIdentifier:@"AddUserVC"];
    vc.delegate = self;
    [self.navigationController pushViewController:vc animated:true];
}


#pragma mark - addNewUser - ListVCDelegate
- (void)addNewUser:(nonnull User *)user {
    [self.items addObject:user];
    [self.tableView reloadData];
}



#pragma mark - UITableViewDataSource, UITextFieldDelegate
- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    ListTableCell *cell =  [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    [cell configure:items[indexPath.row]];
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return items.count;
}


@end

File: WheelsCodingChallenge/View/ListVC.swift
//
//  ListVC.swift
//  WheelsCodingChallenge
//
//  Created by Mike Saradeth on 4/2/19.
//  Copyright © 2019 Mike Saradeth. All rights reserved.
//

import UIKit
import RxSwift
import RxCocoa

class ListVC: UIViewController {
    @IBOutlet weak var tableView: UITableView!
    let disposeBag = DisposeBag()
    var viewModel: ListViewModel!
    
    static func createWith(title: String, viewModel: ListViewModel) -> ListVC {
        let storyboard = UIStoryboard.init(name: "StackUserList", bundle: nil)
        let vc = storyboard.instantiateViewController(withIdentifier: "ListVC") as! ListVC
        vc.title = title
        vc.viewModel = viewModel
        return vc
    }
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupRx()
        viewModel.loadData()
    }
    
    func setupRx() {
        tableView.register(UINib(nibName: "ListTableViewCell", bundle: nil), forCellReuseIdentifier: ListTableViewCell.cellIdentifier)
        
        viewModel.subject.asObservable()
            .bind(to: tableView.rx.items(cellIdentifier: ListTableViewCell.cellIdentifier, cellType: ListTableViewCell.self)) { [weak self]
                (row, item, cell) in
                guard let self = self else { return }
                cell.configure(item: item, row: row, delegate: self.viewModel)
        }
        .disposed(by: disposeBag)
    }

    @IBAction func addUser(_ sender: Any) {
        let badge = Badge(bronze: 0, silver: 0, gold: 0)
        let user = User.init(name: "", urlString: "", reputation: 0, badge: badge)
        let viewModel = AddUserViewModel(user: user)
        let vc = AddUserVC.createWith(title: "Add User", viewModel: viewModel)
        navigationController?.pushViewController(vc, animated: true)
    }
    
}
